File: server/common/protocol.py
from enum import Enum
import logging
from socket import socket
from typing import List

from common.utils import Contestant

BUF_SIZE = 4096
ENDIAN = 'big'
SEPARATOR = ';'
CON_SEPARATOR = '|'
ENCODING = "utf-8"

"""
Response Structure for Winners consulting

| N~WINNERS | PAYLOAD_LENGTH | ID OF WINNERS |
|  2 bytes  |   4 bytes      |  Dynamic      |

"""


class OpCode(Enum):
    CheckClient = 1
    GetStats = 2

    def __str__(self):
        if(self.value == 1):
            return "Check Contestant"

        if(self.value == 2):
            return "Get Stats"


class Protocol:
    def __init__(self, socket: socket) -> None:
        self.socket = socket
        self.__buffer = bytearray()

    def recv_contestants(self) -> Contestant:
        buff_len = self.__recv_message(4)
        value = int.from_bytes(buff_len, ENDIAN)

        buff_parsed = self.__recv_message(value).decode(ENCODING)
        raw_contestants = buff_parsed.split(CON_SEPARATOR)

        contestants = []
        for el in raw_contestants:
            id, first_name, last_name, birth = el.split(SEPARATOR)
            val = Contestant(first_name, last_name, id, birth)

            contestants.append(val)

        return contestants

    def send_response(self, winners: List[Contestant]):

        ids = map(lambda x: x.document, winners)
        ids_buf = bytes(CON_SEPARATOR.join(ids), ENCODING)

        buf = len(ids_buf).to_bytes(4, ENDIAN)
        buf += ids_buf

        self.__send_message(buf)

    def send_partial_stats(self, winners):

        buf = b'1'
        buf += winners.to_bytes(4, ENDIAN)

        self.__send_message(buf)

    def send_definite_stats(self, winners):

        buf = b'0'
        buf += winners.to_bytes(4, ENDIAN)

        self.__send_message(buf)

    def get_next_message_type(self) -> OpCode:
        opcode = self.__recv_message(1)

        value = int.from_bytes(opcode, ENDIAN)
        return OpCode(value)

    def __recv_message(self, buf_len: int) -> List[bytes]:

        while(len(self.__buffer) < buf_len):

            buffer = self.socket.recv(BUF_SIZE)
            if (len(buffer) == 0):
                raise OSError

            self.__buffer += buffer

        val = self.__buffer[:buf_len]
        del self.__buffer[:buf_len]
        return val

    def __send_message(self, buf: bytearray):
        sent = 0
        while(sent < len(buf)):
            sent += self.socket.send(buf[sent:])

    def close(self):
        self.socket.close()

File: client/common/protocol.go
package common

import (
	"bytes"
	"encoding/binary"
	"net"
	"strings"
)

type OpCode byte
const BUFFER = 4096
const SEPARATOR = ';'
const CON_SEPARATOR = '|'

/**
Base Message structure

| Opcode | 
| 1 byte | 

**/

const (
	CheckClient OpCode = 1
)

/**
CheckClient Message Encoded

| Opcode | Length in bytes | Client1 info (ID;FirstName;LastName;)	|
| 1 byte | 	4 bytes 	   |		Dynamic						|

**/

type CheckContestantMessage struct{
	Con []Contestant
	Op OpCode
}

type Protocol struct {
	conn   net.Conn
	read_buffer bytes.Buffer
}

type Contestant struct {
	Name string
	LastName string
	Id string
	Birth string
}

// Encode
func (m *CheckContestantMessage) encode() []byte {
	con := m.Con

	elem := bytes.Buffer{}
	for _, record := range con {
       // Save fields
		elem.WriteString(record.Id)
		elem.WriteByte(SEPARATOR)
		elem.WriteString(record.Name)
		elem.WriteByte(SEPARATOR)
		elem.WriteString(record.LastName)
		elem.WriteByte(SEPARATOR)
		elem.WriteString(record.Birth)
		
		if record != con[len(con)-1] {
			elem.WriteByte(CON_SEPARATOR)
		}
    }

	buf := bytes.Buffer{}
	
	// Store Opcode
	buf.WriteByte(byte(m.Op))

	len_buf := make([]byte, 4)
    binary.BigEndian.PutUint32(len_buf, uint32(elem.Len()))
	buf.Write(len_buf)
	buf.Write(elem.Bytes())
	
	return buf.Bytes()
}

func (c *Protocol) CheckContestant(contestant []Contestant) error {
	message := CheckContestantMessage{
		Op: CheckClient,
		Con: contestant,
	}	
	buf := message.encode()
	
	return c.send(buf)
}

/* Check response for check contestant message. */
func (c *Protocol) CheckResponse() ([]string, error) {

	
	buf, err := c.recv(4)
	if err != nil {
		return nil, err	
	}
    
	length := binary.BigEndian.Uint32(buf)
	
	buf, err = c.recv(int(length))
	
	parsed := string(buf)
	ids := strings.Split(parsed, string(CON_SEPARATOR))
	
	return ids, nil
}

/* Close socket associated to protocol. */
func (c *Protocol) Close() error {
	return c.conn.Close()
}

/* Wrapper for socker send. */
func (c *Protocol) send(buf []byte) error {
	sent_bytes := 0
	for sent_bytes < len(buf) {
		sent, err := c.conn.Write(buf[sent_bytes:])
		sent_bytes += sent
		
		if err != nil {
			return err	
		}
	}
	
	return nil
}

/* Wrapper for socker recv. Internal buffer to avoid short read. */
func (c *Protocol) recv(size int) ([]byte, error) {

	for c.read_buffer.Len() < size {
		buf := make([]byte, BUFFER)
		_, err := c.conn.Read(buf)
		if err != nil {
			return nil, err
		}
		c.read_buffer.Write(buf)
	}

	return c.read_buffer.Next(size), nil
}